### PR TITLE
refactor(protocol): improve Dani's impl

### PR DESCRIPTION
**Pull request by dantaik**
- Remove `state.blockFee`
- Rename `avgFeePerGas` to `feePerGas`
- In bid, keep `blockMaxGasLimit` and init in auction: `newBid.blockMaxGasLimit = config.blockMaxGasLimit;`
- refund proposer correctly with the `feePerGas` value used by the proposer at the proposedAt timestamp, not the value in the auction.
- update `state.proofWindow` correctly
- refactor logic for calculating rewards and update average state vars in block finialzation.
- system prover may prove a block even if there is no auction or auction is ongoing
- fix `hasAuctionEnded`
- fix `pnpm lint:sol` errors
- check `bid.deposit` is large enough
### TODO:
- [ ] create pure function to compare two bids

### CODE REVIEW

File: packages/protocol/contracts/L1/libs/LibAuction.sol [hunk 1]
```diff
@@ -15,127 +15,161 @@ import { TaikoToken } from "../TaikoToken.sol";
...
+        if (
+            bid.prover != address(0) // auto-fill
+                || bid.blockMaxGasLimit != 0 // auto-fill
+                || bid.proofWindow
+                    > state.avgProofTime * config.auctionProofWindowMultiplier
+                || bid.deposit
+                    < state.feePerGas
+                        * (config.blockFeeBaseGas + config.blockMaxGasLimit)
+                        * config.auctionDepositMultipler
+        ) {
```
[the remark this answers is not in the document]
Comment (dantaik) [reply packages/protocol/contracts/L1/libs/LibAuction.sol:43]: If a function is used only once or it's logic is NOT too large , we inline it to reduce cost.

File: packages/protocol/contracts/L1/libs/LibAuction.sol [hunk 2]
```diff
@@ -15,127 +15,161 @@ import { TaikoToken } from "../TaikoToken.sol";
...
+        uint256 count
     )
         internal
-        pure
-        returns (uint64)
+        view
+        returns (uint256 currentTime, TaikoData.Auction[] memory auctions)
     {
-        return 1 + uint64((blockId - 1) / config.auctionBatchSize);
+        if (startBatchId == 0 || count == 0 || count >= 100) {
```
[the remark this answers is not in the document]
Comment (dantaik) [reply packages/protocol/contracts/L1/libs/LibAuction.sol]: They can call this function multiple times to fetch more auctions. I removed the ">=100 check".

File: packages/protocol/contracts/L1/libs/LibAuction.sol [hunk 3]
```diff
@@ -15,127 +15,161 @@ import { TaikoToken } from "../TaikoToken.sol";
...
         }
 
-        return true;
+        TaikoData.Auction memory auction =
+            state.auctions[batchId % config.auctionRingBufferSize];
+
+        if (prover == auction.bid.prover) return true;
+
+        return block.timestamp
+            > auction.startedAt + config.auctionWindow + auction.bid.proofWindow;
```
[the remark this answers is not in the document]
Comment (dantaik) [reply packages/protocol/contracts/L1/libs/LibAuction.sol:148]: We already checked the following in proveBlock, so the block is already proposed and not verified. 

 ````solidity
if (blockId <= state.lastVerifiedBlockId || blockId >= state.numBlocks)
        {
            revert L1_BLOCK_ID();
        }
```

File: packages/protocol/contracts/L1/libs/LibVerifying.sol [hunk 4]
```diff
@@ -182,88 +180,105 @@ library LibVerifying {
...
-                    uint256(auction.startedAt + config.auctionWindowInSec)
-                )
-            );
+        TaikoData.Auction memory auction;
+        {
+            uint64 batchId = LibAuction.batchForBlock(config, blk.blockId);
+            auction = state.auctions[batchId % config.auctionRingBufferSize];
+
+            // this may be false for system prover
+            if (auction.batchId == batchId) {
```
Comment (dantaik) [inline packages/protocol/contracts/L1/libs/LibVerifying.sol]: Changed from 

`if (auction.batchId == batchId) {` to `if (auction.batchId != batchId) {`

[the remark this answers is not in the document]
Comment (dantaik) [reply packages/protocol/contracts/L1/libs/LibVerifying.sol]: Added an assertion. `fc.prover == address(1)` has been checked when the block is proven by system prover.

**Review by adaki2004 — APPROVED** [review]

**Pull request merged**